addons/auth_password_pwn_checker/models/res_users.py

```diff
@@ -0,0 +1,57 @@
+import hashlib
+import logging
+from datetime import timedelta, date
+
+import requests
+
+from odoo import _, models, api, fields
+
+_logger = logging.getLogger(__name__)
+
+
+class ResUsers(models.Model):
+    _inherit = ['res.users']
+
+    pwn_last_checked = fields.Date()
+
+    @classmethod
+    def authenticate(cls, db, login, password, user_agent_env):
+        uid = super(ResUsers, cls).authenticate(db, login, password, user_agent_env)
+        cls._check_user_password(uid, password)
```

Comment:
Also, if I may, as I was pinged.

Is it a desirable behavior to add delay to the login of people?
(Even if at most 3 seconds considering the request timeout).

I would expect this kind of checks to be ran asynchronously, in some kind of delayed CRON maybe.

Reply:
+1, as anyway it sends an email, would delegate to a cron (may use cron triggers for that purpose).

Reply:
Is that true? Can't we store what is sent to the request?

This part:
```py
h = hashlib.sha1(password.encode()).hexdigest().upper()
prefix, suffix = h[:5], h[5:]
```

Reply:
Somehow have it run asynchronous seems like a must, unless we want to immediately redirect to changing password?
If you don't mind the suggestion (even though you probably discussed it before) Couldn't this API be called from the client? Then it could call an endpoint to trigger the email notification, in addition to immediately showing an alert ?

Reply:
I was talking about storing the prefix/suffix, not the whole SHA-1.

If that is also a security concern, which it could be, then we would need to consider another approach.
(Maybe directly sent from the client async as @flch-odoo suggested?).

Reply:
Well if you just need to check the suffix, then it would mean that storing prefix/suffix would be sufficient, and hide everything that is in-between.

Unless I'm really missing something.

(I get that it could *still* be a security concern though, but just making sure I was clear in my suggestion :smile: ).

Reply:
My bad, I thought it was the last 5 characters and not everything but the first 5 :see_no_evil: